File: TempConverter_B4.1/app/src/main/java/com/example/tempconverter/MainActivity.java
package com.example.tempconverter;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    EditText editTextCelsius, editTextFahrenheit;
    Button btnToCelsius, btnToFahrenheit, btnClear;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);  // gắn layout từ res/layout/activity_main.xml

        // Gắn ID từ XML vào biến Java
        editTextCelsius = findViewById(R.id.editTextCelsius);
        editTextFahrenheit = findViewById(R.id.editTextFahrenheit);
        btnToCelsius = findViewById(R.id.btnToCelsius);
        btnToFahrenheit = findViewById(R.id.btnToFahrenheit);
        btnClear = findViewById(R.id.btnClear);

        // Nút chuyển Fahrenheit -> Celsius
        btnToCelsius.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String fStr = editTextFahrenheit.getText().toString();
                if (!fStr.isEmpty()) {
                    try {
                        double f = Double.parseDouble(fStr);
                        double c = (f - 32) * 5 / 9;
                        editTextCelsius.setText(String.format("%.2f", c));
                    } catch (NumberFormatException e) {
                        Toast.makeText(MainActivity.this, "Invalid Fahrenheit input", Toast.LENGTH_SHORT).show();
                    }
                } else {
                    Toast.makeText(MainActivity.this, "Please enter Fahrenheit value", Toast.LENGTH_SHORT).show();
                }
            }
        });

        // Nút chuyển Celsius -> Fahrenheit
        btnToFahrenheit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String cStr = editTextCelsius.getText().toString();
                if (!cStr.isEmpty()) {
                    try {
                        double c = Double.parseDouble(cStr);
                        double f = (c * 9 / 5) + 32;
                        editTextFahrenheit.setText(String.format("%.2f", f));
                    } catch (NumberFormatException e) {
                        Toast.makeText(MainActivity.this, "Invalid Celsius input", Toast.LENGTH_SHORT).show();
                    }
                } else {
                    Toast.makeText(MainActivity.this, "Please enter Celsius value", Toast.LENGTH_SHORT).show();
                }
            }
        });

        // Nút xóa
        btnClear.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                editTextCelsius.setText("");
                editTextFahrenheit.setText("");
            }
        });
    }
}

File: ProgressBar_B20/app/src/main/java/com/example/progressbar/MainActivity.kt
package com.example.progressbar

import android.os.Bundle
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {

    private lateinit var tvPercent: TextView
    private lateinit var progressBar: ProgressBar
    private lateinit var btnStart: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        tvPercent = findViewById(R.id.tvPercent)
        progressBar = findViewById(R.id.progressBar)
        btnStart = findViewById(R.id.btnStart)

        btnStart.setOnClickListener { startProgress() }
    }

    private fun startProgress() {
        progressBar.progress = 0
        tvPercent.text = "0%"

        lifecycleScope.launch {
            Toast.makeText(this@MainActivity, "onPreExecute()", Toast.LENGTH_SHORT).show()

            for (i in 1..100) {
                delay(100)               // giả lập công việc
                progressBar.progress = i
                tvPercent.text = "$i%"
            }

            Toast.makeText(this@MainActivity, "Update xong rồi!", Toast.LENGTH_SHORT).show()
        }
    }
}
